turbopack/crates/turbo-tasks/src/vc/operation.rs

```diff
@@ -0,0 +1,151 @@
+use std::{fmt::Debug, hash::Hash};
+
+use auto_hash_map::AutoSet;
+use serde::{Deserialize, Serialize};
+
+use crate::{trace::TraceRawVcs, CollectiblesSource, RawVc, TaskInput, Upcast, Vc, VcValueTrait};
+
+#[must_use]
+pub struct VcOperation<T>
+where
+    T: ?Sized + Send,
+{
+    pub(crate) node: Vc<T>,
+}
+
+impl<T: ?Sized + Send> VcOperation<T> {
+    /// Creates a new `VcOperation` from a `Vc`.
+    ///
+    /// The caller must ensure that the `Vc` is not a local task and it points to a a single
+    /// operation.
+    pub fn new(node: Vc<T>) -> Self {
+        // TODO to avoid this runtime check, we should mark functions with `(operation)` and return
+        // a VcOperation directly
+        assert!(
+            !node.is_local(),
```

Comment:
Is it valid for a `VcOperation` to be a resolved cell? Does it have to be a function output?

Reply:
This now checks the `RawVc` representation, resolving this issue.